Fix SoftwareSerial support

nobodyinperson opened the pull request:
Compilation failed for me (I use [PlatformIO](https://github.com/platformio)) when using `SoftwareSerial` connections. This fixes it.

The `SoftwareSerial` class obviously does not have these members:

- `flushInput`
- `flushOutput`
- `library_version`

So I commented these calls out to make them no-ops.

Furthermore, somehow `#include`'ing `<SoftwareSerial.h>` in the final Arduino sketch was not enough to trigger the `#ifdef SoftwareSerial_h ... #endif` directives switching `SoftwareSerial` support on in your library. I guess it has to do with [translation unit](https://stackoverflow.com/questions/1106149/what-is-a-translation-unit-in-c) scopes. So I added `#include SoftwareSerial.h` in your library. You probably wanted to keep dependencies low, but it didn't work for me.

My Editor removed trailing whitespace hence the huge amount of modified lines. The real changes are 4 lines IIRC.

Commit:
include <SoftwareSerial.h> and comment unimplemented methods

synfinatic commented on AnySerial.cpp:
```diff
@@ -373,7 +373,7 @@ AnySerial::library_version() {
 #ifdef SoftwareSerial_h
         case anyserial_soft:
-            ret = serialport.soft->library_version();
+            // not implemented
```
I don't have the code in front of me, but looks like there is a static class method to query the version.

nobodyinperson replied:
This might be a problem with me using the super-outdated Arduino IDE 1.0.5 from the Ubuntu repositories. The `SoftwareSerial`-library doesn't seem to have this method in this version.

nobodyinperson replied:
No, the Arduino IDE 1.0.5's `SoftwareSerial` does not include this method:

```bash
locate SoftwareSerial.h | xargs grep -q library_version && echo "found" || echo "not found"                                                                                                                                             
# not found
```

nobodyinperson replied:
Also in the Arduino repository the file https://github.com/arduino/Arduino/blob/master/hardware/arduino/avr/libraries/SoftwareSerial/src/SoftwareSerial.cpp does not mention any `library_version` method.

synfinatic replied:
:( . fair enough.

nobodyinperson replied:
This simple sketch

```c++
#include <SoftwareSerial.h>
#include <AnySerial.h>

SoftwareSerial SoftSerial(10,11);

void setup()
{
  AnySerial *anyserial = new AnySerial(&SoftSerial);
}

void loop() {}
```

fails with the up-to-date Arduino IDE 1.8.5:

```
Arduino: 1.8.5 (Linux), Board: "Arduino/Genuino Mega or Mega 2560, ATmega2560 (Mega 2560)"

/tmp/ccGWKRcD.ltrans0.ltrans.o: In function `main':
ccGWKRcD.ltrans0.o:(.text.startup+0xea): undefined reference to `AnySerial::AnySerial(SoftwareSerial*)'
collect2: error: ld returned 1 exit status
exit status 1
Fehler beim Kompilieren für das Board Arduino/Genuino Mega or Mega 2560.

Dieser Bericht wäre detaillierter, wenn die Option
"Ausführliche Ausgabe während der Kompilierung"
in Datei -> Voreinstellungen aktiviert wäre.
```

Did this work for you previously?

synfinatic commented on AnySerial.h:
```diff
@@ -17,30 +17,31 @@
  */
 
 #include <HardwareSerial.h>
+#include <SoftwareSerial.h>
```
Read the comment above. You need to import SoftwareSerial.h BEFORE AnySerial.h because you may or may not require it and I didn't want to force it.

nobodyinperson replied:
I am `#import <SoftwareSerial.h>`ing **before** `#import <AnySerial.h>`ing, but it keeps saying `undefined reference to 'AnySerial::AnySerial(SoftwareSerial*)'` which AFAIK means the `#ifdef SoftwareSerial_h ... #endif` in your library don't work because the `SoftwareSerial_h`-macro is unset during compilation of your library for whatever reason. In the Arduino IDE 1.0.5 this macro is definitely set in the `SoftwareSerial` library.

Commit:
fix compilation warnings

- remove unused anyserial_none enum value (alternatively handle
  anyserial_none in every switch statement...)
- only define anyserial_* enum values if libraries were included

Commit:
move parts into header

Commit:
move more methods into header

Commit:
add example

nobodyinperson commented:
By moving all the 

```c++
#ifdef SoftwareSerial_h
...
#endif
```

-bits into the header file (which is the one actually included by the sketches) and marking all those methods as `inline` (preventing them to be included multiple times which leads to compilation errors), the `AnySerial` library now works as expected for me.

One can `#include <SoftwareSerial.h>` before `#include <AnySerial.h>` to enable the SoftwareSerial library support. Only then the SoftwareSerial-related parts get compiled in at all.

I also added an example file to point this out.

synfinatic commented:
So I'm running the latest Arduino IDE (1.8.5) and using teensy-dsc as my test (since that's what I built this library for).  With the current master it compiles fine.  With your PR it doesn't:
```
...
linking everything together...
"/Applications/Arduino.app/Contents/Java/hardware/teensy/../tools/arm/bin/arm-none-eabi-gcc" -O2 -Wl,--gc-sections,--relax,--defsym=__rtc_localtime=1525198938 "-T/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/mk20dx256.ld" -lstdc++  -mthumb -mcpu=cortex-m4 -fsingle-precision-constant -o "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/teensy_dsc.ino.elf" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/teensy_dsc.ino.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/Encoder/Encoder.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/AnySerial/AnySerial.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/WiFi_Shield/WiFly.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/WiFi_Shield/WiFlyClient.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/Flash/Flash.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/MsTimer2/MsTimer2.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/EEPROM/EEPROM.cpp.o" "/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/core/core.a" "-L/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475" -larm_cortexM4l_math -lm
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/teensy_dsc.ino.cpp.o: In function `AnySerial::AnySerial(usb_serial_class*)':
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:150: multiple definition of `AnySerial::AnySerial(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:150: first defined here
/Applications/Arduino.app/Contents/Java/hardware/tools/arm/bin/../lib/gcc/arm-none-eabi/5.4.1/../../../../arm-none-eabi/bin/ld: Disabling relaxation: it will not work with multiple definitions
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/teensy_dsc.ino.cpp.o: In function `AnySerial::flush()':
/Users/aturner/Dropbox/Arduino/libraries/Encoder/Encoder.h:813: multiple definition of `AnySerial::AnySerial(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/teensy_dsc.ino.cpp.o: In function `AnySerial::flush()':
/Users/aturner/Dropbox/Arduino/libraries/Encoder/Encoder.h:813: multiple definition of `AnySerial::attach(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/AnySerial/AnySerial.cpp.o: In function `AnySerial::flush()':
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:241: multiple definition of `AnySerial::AnySerial(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/AnySerial/AnySerial.cpp.o: In function `AnySerial::flush()':
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:241: multiple definition of `AnySerial::AnySerial(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/AnySerial/AnySerial.cpp.o: In function `AnySerial::flush()':
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:241: multiple definition of `AnySerial::attach(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/WiFi_Shield/WiFly.cpp.o: In function `AnySerial::AnySerial(usb_serial_class*)':
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:150: multiple definition of `AnySerial::AnySerial(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
Multiple libraries were found for "Encoder.h"/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/WiFi_Shield/WiFly.cpp.o: In function `WiFly::flush()':

 Used: /Users/aturner/Dropbox/Arduino/libraries/Encoder
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:101: multiple definition of `AnySerial::AnySerial(usb_serial_class*)'
 Not used: /Applications/Arduino.app/Contents/Java/hardware/teensy/avr/libraries/Encoder
Multiple libraries were found for "Flash.h"
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
 Used: /Users/aturner/Dropbox/Arduino/libraries/Flash
 Not used: /Users/aturner/Dropbox/Arduino/libraries/arduino_823314/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/WiFi_Shield/WiFly.cpp.o: In function `WiFly::flush()':

/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:101: multiple definition of `AnySerial::attach(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/WiFi_Shield/WiFlyClient.cpp.o: In function `AnySerial::AnySerial(usb_serial_class*)':
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:150: multiple definition of `AnySerial::AnySerial(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/WiFi_Shield/WiFlyClient.cpp.o: In function `AnySerial::flush()':
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:101: multiple definition of `AnySerial::AnySerial(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/libraries/WiFi_Shield/WiFlyClient.cpp.o: In function `AnySerial::flush()':
/Users/aturner/Dropbox/Arduino/libraries/AnySerial/AnySerial.h:101: multiple definition of `AnySerial::attach(usb_serial_class*)'
/var/folders/p3/tcys1lyd4yd1sxkcs63mcq580000gn/T/arduino_build_232475/sketch/cli.cpp.o:/Applications/Arduino.app/Contents/Java/hardware/teensy/avr/cores/teensy3/Print.h:61: first defined here
collect2: error: ld returned 1 exit status
```

nobodyinperson commented:
### teensy-dsc

Alright. In your `teensy-dsc` project you are not using `SoftwareSerial` at all, hence your Anyserial/master-branch just works I guess. But **if** you used a virtual serial port for your `teensy-dsc` WiFly connection in `teensy_dsc.ino` like this:

```diff
 #include <Encoder.h>
+#include <SoftwareSerial.h>
 #include <AnySerial.h>
 #include <WiFly.h>
 #include <Flash.h>
@@ -40,7 +41,8 @@ Encoder EncoderRA(CHAN_A_RA, CHAN_B_RA);
 Encoder EncoderDEC(CHAN_A_DEC, CHAN_B_DEC);
 
 AnySerial UserSerial(&USER_SERIAL_PORT);
-AnySerial WiFlySerialPort(&WIFLY_SERIAL_PORT);
+SoftwareSerial WiFlySoftwareSerialPort(10,11);
+AnySerial WiFlySerialPort(&WiFlySoftwareSerialPort);
 WiFly _WiFly(WiFlySerialPort);

```

... compilation eventually fails (after numerous warnings about unhandled switch-cases) with

```
.../teensy-dsc/src/teensy_dsc/teensy_dsc.ino:45: undefined reference to `AnySerial::AnySerial(SoftwareSerial*)'
``` 

... which means that just using the `SoftwareSerial_h` macro in the `AnySerial.cpp` to elegantly add `SoftwareSerial` support to the `AnySerial` library unfortunately does not work because the macro isn't set globally, only during compilation of the `teensy_dsc.ino` file, not of the `AnySerial.cpp` file.

### multiple definition errors

As for the warnings from my code you mentioned, the compiler seems to choose to ignore the `inline` keywords for the `teensy-dsc`-project. This keyword should actually prevent exactly those errors. For my simple example in the PR it works, pity that it doesn't in more complex scenarios.

### Solutions

So the current situation is not ideal. `AnySerial` does not support `SoftwareSerial` as it is supposed to be and my PR here does not compile reliably.

#### Global compiler flag

One rather unpleasant fix could be to use a (global) compiler flag (like `-Duse_softwareserial`) and to handle that in `AnySerial.{h,cpp}` to properly add `SoftwareSerial` support in the right places. But this would require the library user to tinker with the compilation flags (which I am not really sure whether it's even possible with Arduino IDE...).

#### Force SoftwareSerial inclusion

The simplest solution would be to just do an `#include <SoftwareSerial.h>` in `AnySerial.h`. But this is crude and unnecessarily takes up valuable Flash space when the library user does not need virtual serial ports.

Any further ideas?

Commit:
include <SoftwareSerial.h> in header
